module/zfs/zfs_quota.c

```diff
@@ -350,15 +351,29 @@ zfs_set_userquota(zfsvfs_t *zfsvfs, zfs_userquota_prop_t type,
 		VERIFY(0 == zap_add(zfsvfs->z_os, MASTER_NODE_OBJ,
 		    zfs_userquota_prop_prefixes[type], 8, 1, objp, tx));
 	}
-	mutex_exit(&zfsvfs->z_lock);
 
 	if (quota == 0) {
 		err = zap_remove(zfsvfs->z_os, *objp, buf, tx);
 		if (err == ENOENT)
 			err = 0;
+		/*
+		 * If the quota contains no more entries after the entry
+		 * was removed, destroy the quota zap and remove the
+		 * reference from zfsvfs. This will save us unnecessary
+		 * zap_lookups for the quota during writes.
+		 */
+		VERIFY(0 == zap_count(zfsvfs->z_os, *objp, &zap_nentries));
```

Comment:
```suggestion
		VERIFY0(zap_count(zfsvfs->z_os, *objp, &zap_nentries));
```

Is now preferred since the errno will be logged on failure.  Although as I'm sure you noticed many places in the code haven't yet been updated.   Feel free to update the other callers in this function.